src/flow/persistence/mod.rs

```diff
@@ -253,6 +286,64 @@ impl GroupCommitQueueSet {
             .min()
     }
 
+    /// Retrieves a vector of packets from the persistent log.
+    pub fn retrieve_recovery_packets(&self, nodes: &DomainNodes) -> Vec<Box<Packet>> {
+        let mut packets = vec![];
+        for (_index, node) in nodes.iter() {
+            let node = node.borrow();
+            let local_addr = node.local_addr();
+            let global_addr = node.global_addr();
+            let path = self.log_path(&local_addr);
+            if !path.exists() {
+                continue;
+            }
+
+            let file = File::open(path).unwrap();
+            BufReader::new(file)
+                .lines()
+                .flat_map(|line| {
+                    let line = line.unwrap();
+                    let records: Vec<LogEntry> = serde_json::from_str(&line).unwrap();
+                    records
+                })
+                .enumerate()
+                .map(|(i, entry)| {
+                    let link = Link::new(*local_addr, *local_addr);
+                    let data = entry.records.into_owned();
+                    box match entry.packet_type {
+                        PacketType::Message => Packet::Message {
+                            link,
+                            data,
+                            tracer: None,
+                        },
+                        PacketType::Transaction => {
+                            let id = checktable::TransactionId(i as u64);
+                            let transactions = vec![(id, data.clone(), None)];
+                            let request = checktable::service::TimestampRequest {
+                                transactions,
+                                base: global_addr,
+                            };
+
+                            let reply = self.checktable.apply_batch(request).unwrap().unwrap();
```

Comment:
Not sure if this should handle `None` as well, like in `merge_transactional_packets`. I thought maybe it didn't have to since all the transactions are already committed, but that might be a false assumption.

Reply:
It should not be possible to get back `None` since the packet doesn't have a token (and thus is guaranteed to commit).

However, we may not want to use apply_batch since it is actually a somewhat expensive operation. In this specific case, it should be possible to write something simpler that doesn't painstakingly simulate all the granular checktables since the beginning of time. Plus, then you wouldn't have to deal with the TransactionId stuff.

Reply:
I used `checktable.apply_unconditional` originally - not sure if that's better?